### ISSUE TITLE
Preferred data transfer between JS and WASM context → shouldn't it be WKB?

### ISSUE THREAD

**Issue by chrispahm** [chrispahm/geos-wasm#4]
Currently, the data transfer between the JS and WASM context looks like this:

1) [Converting GeoJSON to WKT in JS](https://github.com/chrispahm/geos-wasm/blob/8c2679536bd33205377b6359a2e7bdc03d69db14/src/allJsFunctions/Buffer.mjs#L182)
2) [Transferring the string to the WASM context](https://github.com/chrispahm/geos-wasm/blob/8c2679536bd33205377b6359a2e7bdc03d69db14/src/allJsFunctions/GeomFromWKT.mjs#L5)
3) Once GEOS is finished, [doing the same thing backwards](https://github.com/chrispahm/geos-wasm/blob/8c2679536bd33205377b6359a2e7bdc03d69db14/src/allJsFunctions/GeomToWKT.mjs)

As @kylebarron wrote in [his blog post](https://kylebarron.dev/blog/geos-wasm), this is highly inefficient! 
And while I totally agree that in the long run, GeoRust + GeoArrow would probably be the preferred options, I still think it would be nice to achieve the "best possible with least amounts of effort" performance with this library in the meantime.

@sanak created a [PR](https://github.com/chrispahm/geos-wasm/pull/3) that makes use of GEOS own GeoJSON Reader/Writer instead of relying on WKT for data exchange. Thanks for that! 🙏 
Using this, we'd be saving the costs of converting formats in JS, but on the other hand serialized GeoJSON can will be very large if there are many properties attached. If we're passing a GeoJSON `FeatureCollection`, we should at least only stringify the `"geometry"`property of each `feature`, as [GEOS won't read any of the `properties` anyway](https://libgeos.org/specifications/geojson/). Again, as @kylebarron [notes](https://kylebarron.dev/blog/geos-wasm#serialization-is-costly), we should avoid serialization as much as possible, which is why I'm not sure if this should be the preferred option.

Another thing Kyle suggested is that using WKB would potentially provide a small speedup, so I started working on that in WKB branch: https://github.com/chrispahm/geos-wasm/tree/wkb.
Using WKB as a data transfer method, the implementation most often beats turf 🚀 
For (very) small and simple features, the speed up is relatively small, and vanishes with increasing number of runs. For large and complex features, using GEOS is almost 29x faster than turf!

```
Testing feature Simple
GEOS buffer method took 88 milliseconds to run 500 times.
Average time per iteration: 0 milliseconds. (lower is better)
@turf/turf buffer method took 105 milliseconds to run 500 times.
Average time per iteration: 0 milliseconds. (lower is better)

Testing feature Australia
GEOS buffer method took 4523 milliseconds to run 5 times.
Average time per iteration: 905 milliseconds. (lower is better)
@turf/turf buffer method took 130323 milliseconds to run 5 times.
Average time per iteration: 26065 milliseconds. (lower is better)
```

**The reason I didn't merge this yet is because it's not consistently working so far.** It usually works for the first run, but subsequent runs most often fail with a `GeometryType X not supported` thrown from the WKX library. X is alway a positive integer not in the range of any "real" geometry types, so I'm quiet sure there's a memory issue at hand, but I can't figure it out yet! Also, I'm having issues polyfilling Buffer with Rollup, as `rollup-plugin-polyfill-node` doesn't seem to notice the use of Buffer in https://github.com/chrispahm/geos-wasm/blob/wkb/src/allJsFunctions/GeomToWKB.mjs

So what do you think, should WKB be the preferred way? If so, I'd really appreciate any help debugging the WKB branch! In any case, let me know what you think!

**Comment by sanak** [comment]
@chrispahm Thanks for sharing about WKB info. (29x faster is really great 😄)

Personally, I prefer that users can choose how to serialize/deserialize from/to GEOS Geometry by themselves, like shapely IO.
https://shapely.readthedocs.io/en/stable/io.html
- from_geojson
- to_geojson
- from_wkb
- to_wkb
- from_wkt
- to_wkt

Also, it would be nice to support direct geometry creation, something like `const point = new Point([0, 0])` or `const point = geos.point([0, 0])`, but it will be next step.
- Shapely: https://shapely.readthedocs.io/en/stable/creation.html
- turf: https://turfjs.org/docs/#point

If you have some rough JS user side API design image, sharing it in a Wiki page or issue is really helpful. 🙇

Note that I am not so familiar with recent JavaScript and WebAssembly is 1st time for me, my contribution pace will be quite slow (also for a weekend work).
But I have some motivation about GEOS + JavaScript wrapper (from past https://osdn.net/projects/geos-com (for Windows IE/ActiveX) to https://github.com/kashif/node-geos (for Node)), and I feel that this GEOS + WASM is the best solution for future, so I am glad to contribute this project. 😊

**Comment by kylebarron** [comment]
> so I'm quiet sure there's a memory issue at hand

I think your issue is here
https://github.com/chrispahm/geos-wasm/blob/97293b919c4855a32dafa8bf71d42b668f518c5a/src/allJsFunctions/GeomToWKB.mjs#L7-L11

When you create a `Uint8Array` there, you're creating a view on the wasm memory directly, without copying. The direct views are really powerful but also really risky. E.g. if the wasm memory gets enlarged, that `Uint8Array` [will become invalid](https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/Memory/grow#detachment_upon_growing).

Especially for these small single geometries, you should be copying them all the time I'd think. You can copy them using [this](https://github.com/kylebarron/arrow-js-ffi/blob/86767702797de47c5351fe63c63d58e56f8493a3/src/vector.ts#L401-L412) for example.

> I'm having issues polyfilling Buffer with Rollup

I think it would be much easier if you used the portable [`ArrayBuffer`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) rather than trying to polyfill the node-specific `Buffer`.

**Comment by chrispahm** [comment]
Thanks @sanak and @kylebarron for your answers!

I'll take it that indeed WKB should probably be preferred for the high-level implementation of `buffer`, but there should be a way to export the low-level functions so users could also roll their own implementations. There's a [new issue now focusing on the API implementation in general](https://github.com/chrispahm/geos-wasm/issues/5), so we can discuss that part there.

Thanks also @kylebarron for debugging the memory issue in the wkb branch! I was following [this web.dev tutorial](https://web.dev/emscripting-a-c-library/#grand-finale-encode-the-image), and missed that critical part where the `Uint8Array` view is copied. Unfortunately, I have to polyfill the Node buffer module since the only [JS WKB parser](https://github.com/syncpoint/wkx) I found solely takes Node buffers as inputs.

Anyways, I published those fixes to the [`wkb` PR](https://github.com/chrispahm/geos-wasm/pull/6), please feel free to have a look! The overall performance is still worse than Turfs, those 29x improvements I found earlier were definitely wrong and probably caused by the memory issues in the geos-wasm package. If there are not obligations, I'll merge the WKB branch and close this issue.

Here's a performance comparison between the current main branch (using WKT) and the WKB feature branch:

## Performance tests
### Performance tests - WKB branch
*Testing feature Simple*
GEOS buffer method took 26 milliseconds to run 100 times.
Average time per iteration: 0 milliseconds.
@turf/turf buffer method took 52 milliseconds to run 100 times.
Average time per iteration: 1 milliseconds.
**GEOS buffer method was 2x faster than @turf/turf buffer method.**

*Testing feature Australia*
GEOS buffer method took 7987 milliseconds to run 100 times.
Average time per iteration: 80 milliseconds.
@turf/turf buffer method took 6633 milliseconds to run 100 times.
Average time per iteration: 66 milliseconds.
**@turf/turf buffer method was 1.2x faster than GEOS buffer method.**

**Real life performance test**
*Buffering all geometries at once*
GEOS buffer method took 315 milliseconds to run 1 times.
Average time per iteration: 315 milliseconds.
@turf/turf buffer method took 177 milliseconds to run 1 times.
Average time per iteration: 177 milliseconds.
**@turf/turf buffer method was 1.78x faster than GEOS buffer method.**

*Buffering each geometry individually*
GEOS buffer method took 278 milliseconds to run 1 times.
Average time per iteration: 278 milliseconds.
@turf/turf buffer method took 135 milliseconds to run 1 times.
Average time per iteration: 135 milliseconds.
**@turf/turf buffer method was 2.06x faster than GEOS buffer method.**

### Performance tests - Main branch (using WKT)
*Testing feature Simple*
GEOS buffer method took 55 milliseconds to run 100 times.
Average time per iteration: 1 milliseconds.
@turf/turf buffer method took 51 milliseconds to run 100 times.
Average time per iteration: 1 milliseconds.
**@turf/turf buffer method was 1.08x faster than GEOS buffer method.**

*Testing feature Australia*
GEOS buffer method took 24946 milliseconds to run 100 times.
Average time per iteration: 249 milliseconds.
@turf/turf buffer method took 6565 milliseconds to run 100 times.
Average time per iteration: 66 milliseconds.
**@turf/turf buffer method was 3.8x faster than GEOS buffer method.**

**Real life performance test**
*Downloads all NUTS3 geometries from Eurostat and buffers them by 1km*
*Buffering all geometries at once*
GEOS buffer method took 1212 milliseconds to run 1 times.
Average time per iteration: 1212 milliseconds.
@turf/turf buffer method took 165 milliseconds to run 1 times.
Average time per iteration: 165 milliseconds.
**@turf/turf buffer method was 7.35x faster than GEOS buffer method.**

*Buffering each geometry individually*
GEOS buffer method took 1174 milliseconds to run 1 times.
Average time per iteration: 1174 milliseconds.
@turf/turf buffer method took 135 milliseconds to run 1 times.
Average time per iteration: 135 milliseconds.
**@turf/turf buffer method was 8.7x faster than GEOS buffer method.**

**Comment by sanak** [comment]
@chrispahm Okay, thanks!
I will take a look at WKB PR (#6) by tomorrow night.

**Comment by kylebarron** [comment]
> Unfortunately, I have to polyfill the Node buffer module since the only [JS WKB parser](https://github.com/syncpoint/wkx) I found solely takes Node buffers as inputs.

You missed the one I wrote 😅 https://github.com/visgl/loaders.gl/blob/master/modules/wkt/src/lib/parse-wkb.ts. Docs are [here](https://github.com/visgl/loaders.gl/blob/7efdbe09e02098aad6d985e4d6465d08806e19a9/docs/modules/wkt/api-reference/wkb-loader.md). It's a little confusing because the `parseWKB` method loads to a custom binary format instead of geojson directly, and then you have to use [`binaryToGeojson`](https://github.com/visgl/loaders.gl/blob/7efdbe09e02098aad6d985e4d6465d08806e19a9/docs/modules/gis/api-reference/binary-to-geojson.md) to convert that binary format to geojson. But it is at least an option to avoid pulling in node dependencies

**Comment by chrispahm** [comment]
Oh, missed that one indeed 😄 I'd update #6 to use your WKB parser instead, but following the discussion in #5 I'm more and more convinced that this repo might as well *just* expose the unaltered GEOS API, and maybe leave the buffer implementation as an example on how it can be used

**Comment by chrispahm** [comment]
Hey @kylebarron, I tried to switch to the loaders.gl WKB parser, and only go to the point where I'm trying to parse the binary format into GeoJSON again: https://github.com/chrispahm/geos-wasm/blob/06198a192be90301bcbe13db8d81eddcf6c32db0/src/allJsFunctions/Buffer.mjs#L222-L223. This fails due to an undefined property `featureIds`, so I gave it a mock value just to see if it would work then https://github.com/chrispahm/geos-wasm/blob/06198a192be90301bcbe13db8d81eddcf6c32db0/src/allJsFunctions/Buffer.mjs#L224-L226. However, the resulting GeoJSON has no coordinates, so I'm doing something wrong here but not sure what it is.
